### PR TITLE
WebGPURenderer: Initial instancing support

### PR DESCRIPTION
**Pull request by takahirox**
This PR adds initial instancing support to WebGPU renderer for `InstancedBufferGeometry` and `InstancedBufferAttribute`. No `InstancedMesh` support yet.

Basic changes are

* If geometry is `InstancedBufferGeometry` (and `.instanceCount` is non-zero), the second argument of `passEncoder.draw/.drawIndexed` is set to `InstancedBufferGeometry.instanceCount`. Otherwise set to 1.
* If attribute is `InstancedBufferAttribute`, corresponding stepMode of vertex buffer layout descriptor is set to `'instance'`. Otherwise set to `'vertex'`.

Demo computing instance 128k box positions in compute shader and rendering them with instancing: 

https://raw.githack.com/takahirox/three.js/WebGPUInstancingTrial/examples/index.html#webgpu_instancing

image

Shader code in demo: 

https://github.com/takahirox/three.js/blob/ddbb33e22d4acb73befa32f4abe488b509744baf/examples/webgpu_instancing.html#L151-L188

Note that this example is not in this PR because it requires custom shader but I don't want to change or add built-in shader yet until node based material system lands.

So I made another branch for the demo. Also the branch includes [this change](https://github.com/mrdoob/three.js/pull/20319#issuecomment-697166584) for custom shader.

### CODE REVIEW

**Comment by Mugen87** [comment]
I had this also implemented on my local computer but I've not pushed it since it was not sure @mrdoob wants to support `InstancedBufferGeometry`. When I remember correctly, there were some considerations about the API. Sorry, if I'm wrong here.

@mrdoob Still okay to merge this PR?

File: examples/jsm/renderers/webgpu/WebGPURenderer.js [hunk 1]
```diff
@@ -780,7 +780,7 @@ class WebGPURenderer {
 			const positionAttribute = geometry.attributes.position;
 			const vertexCount = ( drawRange.count !== Infinity ) ? drawRange.count : positionAttribute.count;
 
-			passEncoder.draw( vertexCount, 1, firstVertex, 0 );
+			passEncoder.draw( vertexCount, geometry.instanceCount || 1, firstVertex, 0 );
 
 			info.update( object, vertexCount );
```
Comment (Mugen87) [inline examples/jsm/renderers/webgpu/WebGPURenderer.js:785]: It's necessary to update `WebGPUInfo` to report correct number of rendered primitives.

**Comment by Mugen87** [comment]
@Kangz In WebGL, it was necessary to use `vertexAttribDivisor()` how often  a value of the instanced attribute should be repeated. E.g. a divisor of one means that each value of the instanced attribute is used for a single instance. A divisor of two would mean that each value is used for two consecutive instances. `three.js` allows to controls setting via `InstancedBufferAttribute.meshPerAttribute`.

Is there a counterpart for this in WebGPU? Or is the divisor considered to be always 1?

**Comment by Kangz** [comment]
Unfortunately the divisor is always 1. This is a constraint that comes from Vulkan where having a variable rate divisor is an optional feature (that's largely not present on mobile).

**Comment by mrdoob** [comment]
> I had this also implemented on my local computer but I've not pushed it since it was not sure @mrdoob wants to support `InstancedBufferGeometry`. 

I don't know what are the benefits of `InstancedBufferGeometry` vs `InstancedMesh` 🤔

**Comment by Mugen87** [comment]
The advantage of `InstancedBufferGeometry` is that it is more flexible. For example you can use it to render instanced lines.

On the other hand, one can consider  `InstancedBufferGeometry` as very low-level and thus hard to use especially for beginners. Without knowing shader programming, you can't really use this class.

**Comment by takahirox** [comment]
I thought the same thing, "Do we really need `InstancedBufferGeometry`?".

With node based material system, writing custom shader will be easier. Then I guess some users may like to have not only instanced matrix/color attributes but also their user defined instance attributes. So I think we need `InstancedBufferGeometry`, or we may need to add more flexibility to `InstancedMesh`.

**Comment by Mugen87** [comment]
> With node based material system, writing custom shader will be easier.

Yeah, that makes me hope that `InstancedBufferGeometry` will be easier to use. The main reasons for introducing `InstancedMesh` was to not confront the user with writing/patching shaders. If materials will be composed differently in the future, the entire class might become obsolete.

Besides, the changes to the renderer are straightforward. The PR highlights the critical parts which are necessary to support instanced rendering. Meaning configuring `stepMode` and `instanceCount`. The engine might derive these information differently later on but in the current situation it's a good start.

Hence, I vote to add support for `InstancedBufferGeometry` for now.

**Comment by takahirox** [comment]
> The PR highlights the critical parts which are necessary to support instanced rendering. Meaning configuring stepMode and instanceCount. The engine might derive these information differently later on but in the current situation it's a good start.

+1

File: examples/jsm/renderers/webgpu/WebGPURenderer.js (hunk 1 above)
Comment (takahirox) [reply examples/jsm/renderers/webgpu/WebGPURenderer.js]: Thanks, updated. Please double check.

**Review by Mugen87 — APPROVED** [review]

**Comment by voland62** [comment]
Interesting, would it be some benefits using raytrace-accelerated graphics (like Nvidia RTX 3080) via WEbGPURenderer for doing distance field rendering with raytracing instead of shadertoy's-like raymarching?...

**Comment by takahirox** [comment]
@Mugen87 You approved but didn't merge. It means you are waiting for @mrdoob's comment, doesn't it?

**Comment by Mugen87** [comment]
Yep 😇 .

**Comment by mrdoob** [comment]
> The advantage of `InstancedBufferGeometry` is that it is more flexible. For example you can use it to render instanced line segments.

Okay, lets support this for now then 👍

**Pull request merged**